_posts/2020-11-12-deconvolution.md

```diff
@@ -0,0 +1,436 @@
+
+  <!-- Vertical lines -->
+  <line x1="42" y1="32" x2="42" y2="152" style="stroke-width:2" />
+  <line x1="72" y1="32" x2="72" y2="152" />
+  <line x1="102" y1="32" x2="102" y2="152" />
+  <line x1="132" y1="32" x2="132" y2="152" />
+  <line x1="162" y1="32" x2="162" y2="152" style="stroke-width:2" />
+
+  <!-- Colored Rectangle -->
+  <polygon points="42.74356617647059,32.74356617647059 162.74356617647058,32.74356617647059 162.74356617647058,152.74356617647058 42.74356617647059,152.74356617647058" style="fill:#ECB172A0;stroke-width:0"/>
+
+  <!-- Text -->
+  <text x="102.743566" y="172.743566" font-size="1.0rem" font-weight="100" text-anchor="middle" >2048</text>
+  <text x="182.743566" y="92.743566" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,182.743566,92.743566)">2048</text>
+  <text x="16.371783" y="156.371783" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(45,16.371783,156.371783)">950</text>
+</svg>
+</td>
+</tr>
+</table>
+
+
+What we now have is a Dask array composed of 16 CuPy blocks of data.  Notice
+how Dask provides nice typing information in the SVG output.  When we moved
+from NumPy to CuPy, the block diagram above displays `Type: cupy.ndarray` --
+this is a nice sanity check.
+
+The last piece we need before running the deconvolution is the PSF which should
+also be loaded onto the GPU:
+
+```python
+psf = skimage.io.imread("/public/NVMICROSCOPY/PSF.tif")
+c_psf = cp.asarray(psf)
+```
+
+Lastly, we call `map_overlap` with the `deconvolve` function across the Dask
+array:
+
+```python
+out = da.map_overlap(
+    deconvolve,
+    c_imgs,
+    psf=c_psf,
+    iterations=20,
+    meta=c_imgs._meta,
+    depth=tuple(np.array(c_psf.shape) // 2),
+    boundary="periodic"
+)
+out
+```
+
+<a href="/images/deconvolve.png">
+    <img src="/images/deconvolve.png" width="100%"></a>
+
+The image above is taken from a mouse intestine.
+
+With Dask and multiple GPUs, we measured deconvolution of an 8GB image in ~10
+seconds! But this is just the first step towards accelerated image science.
+
+Napari
+------
+
+Deconvolution is just one operation and one tool, an image scientist or
+microscopist will need.  They will need other tools as they study the
+underlying biology.  Before getting to those next steps, they will need tools
+to visualize the data. [Napari](https://napari.org/), a multi-dimensional image
+viewer used in the PyData Bio ecosystem, is a good tool for visualizing this
+data.  As an experiment, we ran the same workflow on a local workstation with 2
+Quadro RTX 8000 GPUs connected with NVLink.  [Example
+Notebook](https://gist.github.com/quasiben/02b3dabba8fb3415e40e685b3cb2ca4a)
+
+By adding a `map_blocks` call to our array, we can move our data _back_ from
+GPU to CPU (device to host).
+
+```python
+def cupy_to_numpy(x):
+    import cupy as cp
+    return cp.asnumpy(x)
+
+np_out = out.map_blocks(cupy_to_numpy, meta=out)
```

Comment:
```suggestion
np_out = out.map_blocks(cp.asnumpy, meta=out)
```

why not just pass cp.asnumpy directly to map_blocks?

Reply:
@quasiben, do you have thoughts here?